src/V4Router.sol

```diff
@@ -0,0 +1,179 @@
+                amountOut =
+                    uint128(_swap(poolKey, zeroForOne, int256(-int128(params.amountIn)), 0, params.path[i].hookData));
+
+                params.amountIn = amountOut;
+                params.currencyIn = params.path[i].intermediateCurrency;
+            }
+
+            if (amountOut < params.amountOutMinimum) revert TooLittleReceived();
+        }
+    }
+
+    function _swapExactOutputSingle(IV4Router.ExactOutputSingleParams memory params) private {
+        _swap(
+            params.poolKey,
+            params.zeroForOne,
+            int256(int128(params.amountOut)),
+            params.sqrtPriceLimitX96,
+            params.hookData
+        );
+    }
+
+    function _swapExactOutput(IV4Router.ExactOutputParams memory params) private {
+        unchecked {
+            uint256 pathLength = params.path.length;
+            uint128 amountIn;
+
+            for (uint256 i = pathLength; i > 0; i--) {
+                (PoolKey memory poolKey, bool oneForZero) =
+                    _getPoolAndSwapDirection(params.path[i - 1], params.currencyOut);
+                amountIn = uint128(
+                    -_swap(poolKey, !oneForZero, int256(int128(params.amountOut)), 0, params.path[i - 1].hookData)
+                );
+
+                params.amountOut = amountIn;
+                params.currencyOut = params.path[i - 1].intermediateCurrency;
+            }
+            if (amountIn > params.amountInMaximum) revert TooMuchRequested();
+        }
+    }
+
+    function _swap(
+        PoolKey memory poolKey,
+        bool zeroForOne,
+        int256 amountSpecified,
+        uint160 sqrtPriceLimitX96,
+        bytes memory hookData
+    ) private returns (int128 reciprocalAmount) {
+        BalanceDelta delta = poolManager.swap(
+            poolKey,
+            IPoolManager.SwapParams(
+                zeroForOne,
+                amountSpecified,
+                sqrtPriceLimitX96 == 0
+                    ? (zeroForOne ? TickMath.MIN_SQRT_PRICE + 1 : TickMath.MAX_SQRT_PRICE - 1)
+                    : sqrtPriceLimitX96
+            ),
+            hookData
+        );
+
+        reciprocalAmount = (zeroForOne == amountSpecified < 0) ? delta.amount1() : delta.amount0();
+    }
+
+    function _getPoolAndSwapDirection(PathKey memory params, Currency currencyIn)
+        private
+        pure
+        returns (PoolKey memory poolKey, bool zeroForOne)
+    {
+        (Currency currency0, Currency currency1) = currencyIn < params.intermediateCurrency
+            ? (currencyIn, params.intermediateCurrency)
+            : (params.intermediateCurrency, currencyIn);
+
+        zeroForOne = currencyIn == currency0;
+        poolKey = PoolKey(currency0, currency1, params.fee, params.tickSpacing, params.hooks);
+    }
+
+    function _take(Currency currency, address recipient) private {
+        int256 delta = poolManager.currencyDelta(address(this), currency);
+        if (delta < 0) revert();
```

Comment:
custom error?

Reply:
also I feel like TAKE should accept a parameter? not just use the open delta? what if you want to take then swap then pay back or something?

Reply:
also todo.. use CurrencySettle library

Reply:
> what if you want to take then swap then pay back or something?

what do you mean by that?
but yes, either way, when it inherits from DeltaResolver it will take an amount field

Reply:
>what do you mean by that?

i understand this comment now. Yeah this will be done when `DeltaResolver` is inherited 👯